src/main/java/org/spdx/utility/compare/TemplateRegexGenerator.java

```diff
@@ -0,0 +1,322 @@
+
+	/**
+	 * Generates regular expressions from a license or exception template
+	 * @throws SpdxCompareException 
+	 */
+	public TemplateRegexGenerator(String template) throws SpdxCompareException {
+		this.template = template;
+		parseTemplate();
+	}
+	
+	/**
+	 * Parses the template generating the regular expression
+	 * @throws SpdxCompareException 
+	 */
+	private void parseTemplate() throws SpdxCompareException {
+		try {
+			SpdxLicenseTemplateHelper.parseTemplate(LicenseCompareHelper.removeCommentChars(template), this);
+		} catch (LicenseTemplateRuleException e) {
+			throw new SpdxCompareException("Invalid template rule found during filter: "+e.getMessage(),e);
+		} catch (LicenseParserException e) {
+			throw new SpdxCompareException("Invalid template found during filter: "+e.getMessage(),e);
+		}
+	}
+	
+	/**
+	 * @return the complete regular expression for the template
+	 */
+	public String getCompleteRegex() {
+		return regexPatternList.toString();
+	}
+	
+	/**
+	 * @param wordLimit number of non optional words to include in the pattern
+	 * @return a regex to match the start of the license per the template
+	 */
+	public String getStartRegex(int wordLimit) {
+		RegexList result = new RegexList();
+		int index = 0;
+		int numWords = 0;
+		List<RegexElement> elementList = regexPatternList.getElements();
+		while (index < elementList.size() && numWords <= wordLimit) {
+			RegexElement element = elementList.get(index++);
+			result.addElement(element);
+			if (element instanceof RegexToken) {
+				numWords++;
+			}
+		}
+		// Need to check if the string starts with a greedy regex and change to non-greedy
+		if (result.getElements().size() > 0) {
+			RegexElement firstElement = result.getElements().get(0);
+			if (firstElement instanceof RegexPattern) {
+				String pattern = ((RegexPattern)firstElement).pattern;
+				if (!pattern.startsWith(".?") && pattern.startsWith(".")) {
+					((RegexPattern)firstElement).setPattern(".?" + pattern.substring(1));
+				}
+			}
+		}
+		return result.toString();
+	}
+	
+	/**
+	 * @param wordLimit number of non optional words to include in the pattern
+	 * @return a regex to match the end of the license per the template
+	 */
+	public String getEndRegex(int wordLimit) {
+		RegexList result = new RegexList();
+		int numWords = 0;
+		List<RegexElement> elementList = regexPatternList.getElements();
+		int index = elementList.size()-1;
+		// find the beginning of the end ...
+		while (index > 0 && numWords <= wordLimit) {
+			if (elementList.get(index--) instanceof RegexToken) {
+				numWords++;
+			}
+		}
+		while (index < elementList.size()) {
+			result.addElement(elementList.get(index++));
+		}
+		return result.toString();
+	}
+
+	/**
+	 * @param text
+	 * @return the text matching the beginning and end regular expressions for the template.  Null if there is no match.
+	 * @throws SpdxCompareException
+	 */
+	public @Nullable String findTemplateWithinText(String text) throws SpdxCompareException {
```

Comment:
Might this be private?  Or are there cases where callers (e.g. `LicenseCompareHelper`) actually need to be able to pull out sub-texts that match a template from within a larger text?

Reply:
This one is used by the `LicenseCompareHelper`.  If we want to make this class more of the main interface, we can make this one private and create a public `isTemplateTextWithin` method.

Reply:
Makes sense.  Of the various methods I thought might be able to be made private, this was the one I was least sure about, as it does seem useful to callers.